### PR TITLE
Adding the new feature of FPDT

### PR DESCRIPTION
**Pull request by YJHMITWEB**
[FPDT](https://arxiv.org/abs/2408.16978) can only be work with [this](https://github.com/microsoft/DeepSpeed/pull/6462) version of DeepSpeed.

### CODE REVIEW

**Comment by tohtana** [comment]
@YJHMITWEB Do we need changes in `gpt2-merge.txt` / `gpt2-vocab.json`?

File: megatron/initialize.py [hunk 1]
```diff
@@ -349,9 +349,12 @@ def _warmup_jit_function():
         dtype = torch.float32
 
     # Warmup fused bias+gelu
+    seq_length = args.seq_length
+    if args.ds_sequence_parallel_fpdt:
+        seq_length = 8192
```
Comment (tohtana) [inline megatron/initialize.py:354]: Can you define this as another variable like "FPDT_SEQ_LEN" and give a description in a comment why we have this setting?

File: megatron/model/rotary_pos_embedding.py [hunk 2]
```diff
@@ -32,7 +35,9 @@ def forward(self, max_seq_len, offset=0):
         emb = torch.cat((freqs, freqs), dim=-1)
         # emb [seq_length, .., dim]
         from einops import rearrange
-        return rearrange(emb, 'n d -> n 1 1 d')
+        base = rearrange(emb, 'n d -> n 1 1 d')
```
Comment (inkcherry) [inline megatron/model/rotary_pos_embedding.py:38]: will this change the output when use ```--use-rotary-position-embeddings```,  in  llama style model, 
 FYI https://github.com/microsoft/Megatron-DeepSpeed/blob/main/examples_deepspeed/pretrain_llama2_distributed.sh?

Comment (YJHMITWEB) [reply megatron/model/rotary_pos_embedding.py:38]: We have tested both GPT and Llama models, this works well with both.

File: megatron/initialize.py (hunk 1 above)
Comment (YJHMITWEB) [reply megatron/initialize.py]: This is fixed by setting it to be `ds_sequence_parallel_fpdt_chunk_size` if FPDT is enabled.

**Comment by YJHMITWEB** [comment]
> > Hi @YJHMITWEB , is FPDT referring to this paper? https://ui.adsabs.harvard.edu/abs/2023JARS...17b6510H/abstract
> 
> @delock, no, FPDT refers to [this paper](https://arxiv.org/abs/2408.16978), aka Ulysses-Offload

Thanks @samadejacobs for pointing.

**Comment by YJHMITWEB** [comment]
> @microsoft-github-policy-service agree

**Comment by YJHMITWEB** [comment]
@microsoft-github-policy-service agree

File: megatron/model/transformer.py [hunk 3]
```diff
@@ -707,21 +721,22 @@ def forward(self, hidden_states, attention_mask,
...
-             value_layer) = self.split_tensor(mixed_x_layer)
+            
+            if self.enable_ds_sequence_parallel:
+                assert self.projection_size == self.kv_projection_size
+                seq_len, bs = mixed_x_layer.shape[0], mixed_x_layer.shape[1]
+                query_layer = mixed_x_layer[:, :, :self.projection_size].reshape(seq_len, bs, -1, self.head_dim)
+                key_layer = mixed_x_layer[:, :, self.projection_size:self.projection_size+self.kv_projection_size].reshape(seq_len, bs, -1, self.head_dim)
+                value_layer = mixed_x_layer[:, :, self.projection_size+self.kv_projection_size:].reshape(seq_len, bs, -1, self.head_dim)
+            if self.sequence_parallel or not self.enable_ds_sequence_parallel:
+                seq_len, bs = mixed_x_layer.shape[0], mixed_x_layer.shape[1]
```
Comment (samadejacobs) [inline megatron/model/transformer.py:735]: Might be worthwhile to keep "split_tensor" implementation as default, that is not fpdt scenario.

**Review by samadejacobs — APPROVED** [review]

**Pull request merged**